python/ray/tests/kuberay/test_autoscaling_e2e.py

```diff
@@ -0,0 +1,290 @@
+        - One Ray head pod
+            - Autoscaler: .25 CPU, .5 Gi memory
+            - Ray node: .5 CPU, .5 Gi memeory
+        - Two Worker pods
+            - Ray node: .5 CPU, .5 Gi memory
+        Total: 1.75 CPU, 2 Gi memory.
+
+        Including operator and system pods, the total CPU requested is around 3.
+
+        The cpu LIMIT of each Ray container is 1.
+        The `num-cpus` arg to Ray start is 1 for each Ray container; thus Ray accounts
+        1 CPU for each Ray node in the test.
+        """
+        # Cluster-creation
+        logger.info("Creating a RayCluster with no worker pods.")
+        self._apply_ray_cr(min_replicas=0, replicas=0)
+
+        logger.info("Confirming presence of head.")
+        wait_for_pods(goal_num_pods=1, namespace="default")
+        head_pod = get_pod(
+            pod_name_filter="raycluster-complete-head", namespace="default"
+        )
+
+        logger.info("Waiting for head pod to start Running.")
+        wait_for_pod_to_start(head_pod, namespace="default")
+        logger.info("Confirming Ray is up on the head pod.")
+        wait_for_ray_health(head_pod, namespace="default")
+
+        # Scale-up
+        logger.info("Scaling up to one worker via Ray resource request.")
+        scale_script = (
+            "import ray;"
+            'ray.init("auto");'
+            "ray.autoscaler.sdk.request_resources(num_cpus=2)"
+        )
+        # The request for 2 cpus should give us a 1-cpu head (already present) and a
+        # 1-cpu worker (will await scale-up).
+        kubectl_exec(
+            command=["python", "-c", scale_script],
+            pod=head_pod,
+            container="ray-head",
+            namespace="default",
+        )
+        logger.info("Confirming number of workers.")
+        wait_for_pods(goal_num_pods=2, namespace="default")
+
+        logger.info("Scaling up to two workers by editing minReplicas.")
+        # replicas=1 reflects the current number of workers
+        # (which is what we expect to be already present in the Ray CR)
+        self._apply_ray_cr(
+            min_replicas=2,
+            replicas=1,
+            # Validate that replicas set on the Ray CR by the autoscaler
+            # is indeed 1:
+            validate_replicas=True,
+        )
+        logger.info("Confirming number of workers.")
+        wait_for_pods(goal_num_pods=3, namespace="default")
+
+        # Scale-down
+        logger.info("Removing resource request.")
+        scale_down_script = (
+            "import ray;"
+            'ray.init("auto");'
+            "ray.autoscaler.sdk.request_resources(num_cpus=0)"
+        )
+        kubectl_exec(
+            command=["python", "-c", scale_down_script],
+            pod=head_pod,
+            container="ray-head",
+            namespace="default",
+        )
+        logger.info("Scaling down all workers by editing maxReplicas.")
+        # (replicas=2 reflects the current number of workers)
+        self._apply_ray_cr(
```

Comment:
It is a bit odd that we are doing this immediately after running the scale down script -- shouldn't the scale down script be enough to scale down? And we should wait for scale down immediately after the scale down script ran.

It is also valueable to test that setting the max_replicase to 0 will scale down, but it seems to me that that should be done independently.

Feel free to address in a separate PR if this one is almost ready to be merged.

Reply:
This is a great point -- scaling down via request resources should allow the autoscaler to scale down, without touching max replicas.

The catch is that it would take 5 minutes because that's the default and the CRD does not currently expose that setting.

I think we should expose that setting, set timeout low for the test, and then test scale-down without the CRD change.

I am actually not sure what the autoscaler does if max workers is set to 0 while a resource request persists, but I don't want to explore that right this second.